### PR TITLE
Updating AppleCider models to work with `to_onnx` and `engine`

### PR DESCRIPTION
**Pull request by drewoldag**
Primarily what is required here is to update the return values from the models `to_tensor` functions to be tuples of numpy arrays, that include empty arrays when running inference. 

Other specific changes:
* Changes to photo_dataset
  * Moved the `collate` function into the class
  * Removed the logic to add extra CLS padding from the pad_mask array. (This now happens in `BaselineCLS.forward`.
* Changes to image_and_metadata_dataset
  * Add toggle for oversampling mixin
* Changes to BaselineCLS
  * `to_tensor` returns a tuple of numpy arrays
  * Removed `pad` input from `forward`
  * Add extra column to pad_mask for CLS padding
* Changes to Spectranet
  * `to_tensor` returns a tuple of numpy arrays
  * `forward` always expects a tuple as input and will expand the tuple with `data, _, _ = x`. 
* Changes to AstroMiNN
  * `to_tensor` returns a tuple of numpy arrays

### CODE REVIEW

File: src/applecider/models/HyraxBaselineCLS.py [hunk 1]
```diff
@@ -95,7 +94,7 @@ def train_step(self, batch):
             Dictionary containing the loss value for the current batch.
         """
 
-        labels = batch[1]
+        _, _, labels = batch
```
Comment (maxwest-uw) [inline src/applecider/models/HyraxBaselineCLS.py]: same q as above, should this be `_, labels, _ = batch` ?

File: src/applecider/datasets/photo_dataset.py [hunk 2]
```diff
@@ -115,29 +115,38 @@ def __len__(self):
...
-             mask], dim=1
-        )
-    object_ids = [b["object_id"] for b in batch]
-    return {"data": {"photometry": pad,
-                     "label": torch.tensor(labels),
-                     "pad_mask": pad_mask,
-                     "mean": torch.tensor(batch[0]["data"]["mean"]),
-                     "std": torch.tensor(batch[0]["data"]["std"]),
-                     },
-            "object_id": object_ids,
```
Comment (dougbrn) [inline src/applecider/datasets/photo_dataset.py]: why don't we need object_id anymore? I recall that that was added for compliance with inference, has hyrax changed for this?

**Review by dougbrn — COMMENTED** [review]
I took a look at the baselineCLS sections and the changes seem about what I'd expect, just had a few questions. Also, do you expect any changes needed to the run script or runtime config to get this working (or alternatively, have you tried to run the run script with these changes)

File: src/applecider/models/HyraxBaselineCLS.py (hunk 1 above)
Comment (drewoldag) [reply src/applecider/models/HyraxBaselineCLS.py]: Same comment as above :) I adjusted the order to help out with onnx-ification. Hopefully in the long run, the ordering won't matter, but it seems to make a difference for now because the to_onnx script will iterate through the tuple of inputs, taking particular actions for those that have data vs. those that are empty. And in this case, during inference, there was an empty array (`labels`) that was confusing the onnx exporter. Moving labels to be the last array in the list resolved the confusion.

File: src/applecider/models/spectranet.py [hunk 3]
```diff
@@ -199,11 +198,16 @@ def train_step(self, batch):
...
+
+        data = data_dict["data"]
         return (
-            torch.tensor(data_dict["data"]["flux"]).to(torch.float32),
-            torch.tensor(data_dict["data"]["label"]).to(torch.int16),
-            torch.tensor(data_dict["data"]["redshift"]).to(torch.float32),
+            np.asarray(data.get("flux", []), dtype=np.float32),
+            # Return empty arrays even if the labels/redshifts are not present
+            np.asarray(data.get("label", []), dtype=np.int16),
+            np.asarray(data.get("redshift", []), dtype=np.float32),
```
[the remark this answers is not in the document]
Comment (drewoldag) [reply src/applecider/models/spectranet.py]: We're going to expect empty arrays to be returned when labels or redshift are not present.
